src/Corax/Queries/SpatialMatch/SpatialMatch.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            _startsWithDisposeHandler = Slice.From(_allocator, _geohashsTermMatch[_geoTermMatchId], out _startsWith);
+            _currentMatch = _indexSearcher.TermQuery(_fieldName, _startsWith);
+            
+            _geoTermMatchId++;
+
+            return true;
+        }
+
+        if (_isTermMatch)
+        {
+            //We are out of TermMatches. Now we are gonna do SEEK and check every single element;
+            _isTermMatch = false;
+
+            if (_geohashsNeedsToBeChecked.Count == 0)
+            {
+                return false;
+            }
+
+            StartsWithFetcher();
+        }
+
+
+        while (true)
+        {
+            //   _currentMatch.
+            if (_startsWithProvider.Next(out _currentMatch, out _startsWith) == false || requireReload)
+            {
+                if (_geohashsNeedsToBeChecked.Count == 0 || _geohashsNeedsToBeChecked.Count <= _geoStartsWithId)
+                {
+                    return false;
+                }
+
+                StartsWithFetcher();
+            }
+            else
+            {
+                _trie.TryGetValue(_startsWith, out var alreadyReturned);
+                if (alreadyReturned)
+                    continue;
+                
+                return true;
+            }
+        }
+
+
+        void StartsWithFetcher()
+        {
+            //_startsWithDisposeHandler?.Dispose();
+            _startsWithDisposeHandler = Slice.From(_allocator, _geohashsNeedsToBeChecked[_geoStartsWithId], out _startsWith);
+            _startsWithProvider = new(_indexSearcher, _allocator, _tree, _fieldName, _fieldId, _startsWith);
+            _geoStartsWithId++;
+        }
+    }
+
+    public long Count => throw new NotSupportedException();
+    public QueryCountConfidence Confidence => QueryCountConfidence.Low;
+    public bool IsBoosting { get; }
+
+    public int Fill(Span<long> matches)
+    {
+        int currentIdx = 0;
+        do
+        {
+            int read;
+            if ((read = _currentMatch.Fill(matches.Slice(currentIdx))) == 0)
+            {
+                if (GoNextMatch() == false)
+                {
+                    return currentIdx;
+                }
+                continue;
+            }
+
+            if (_isTermMatch)
+            {
+                currentIdx += read;
+            }
+            else
+            {
+                var figure = Spatial4n.Core.Util.GeohashUtils.DecodeBoundary(_startsWith.ToString(), _spatialContext);
+                if (IsTrue(figure.Relate(_shape)))
+                {
+                    _trie.Add(_startsWith, true);
+                    GoNextMatch(true);
+                    currentIdx += read;
+                }
+            }
+        } while (currentIdx != matches.Length);
+
+        return currentIdx;
+    }
+
+    public bool IsTrue(Spatial4n.Core.Shapes.SpatialRelation answer) => answer switch
```

Comment:
Let's give it a better name - `MatchRequireRelation` might be better, and the do the `Relate` call internally